### PR TITLE
test(ai-assist): antagonist phase 2 — provider-shape torture tests + Gemini tool-conflict guard

### PR DESCRIPTION
**Pull request by ErikFortune**
## Summary

Phase 2 of the antagonist effort (phase 1 = `@fgv/ts-agent-memory`, #528). Same hole-driven charter pointed at `@fgv/ts-extras/ai-assist` — the provider-shape / model-routing / convert-symmetry / param-rejection near-miss classes drawn from the #520, #522, and #524 near-misses.

**Result: no bugs in classes 1–6 (the corrected provider surface held), and one real pre-existing gap found in class 7 — now fixed** (principal-approved).

## Class 7 — the gap, and the fix

**Finding:** Gemini's `generateContent` API HTTP-400s (`INVALID_ARGUMENT`) when built-in grounding (`web_search`) and function calling (client tools) are combined in one request, and ai-assist had **no pre-flight guard** — the combination serialized straight to the wire, surfacing as an opaque 400.

**Fix:** `executeClientToolTurn` now fails fast with a clear, actionable message when a Gemini turn combines a `web_search` server tool with client tools. The conflict rule is a pure, directly-unit-testable predicate (`hasGeminiToolConflict`, `@internal` — not added to the public API). Other providers accept the mix and are unaffected. The class-7 torture test now asserts the guard (every predicate branch + end-to-end fail-fast) instead of documenting the gap.

## Torture tests, per class (all genuinely adversarial — each names the wrong impl it guards)

1. **finishReason decline-vs-benign** (`apiClient.antagonist.test.ts`) — RECITATION/PROHIBITED_CONTENT/OTHER/empty/unknown-future values → declined; near-miss casing (`stop`/`Stop`/`STOPPED`/`MAX_TOKEN`) NOT treated as benign; completion `truncated` exact-case matrix. Guards "any finishReason = decline", "only SAFETY = decline", case-insensitive/truthy benign checks.
2. **Model routing** (`modelRouting.antagonist.test.ts`) — `isResponsesOnlyModel` substring-not-prefix, case sensitivity, multi-prefix ordering. Guards `.includes()`-instead-of-`.startsWith()`, case-folding, first-prefix-only.
3. **Convert/validate symmetry** (`converters.antagonist.test.ts`) — full-optionals round-trip + absent-stays-absent for every non-enum exported converter (incl. `aiClientToolConfig` with all annotation hints — the #524 class). Guards a field map omitting an interface field.
4. **thinking↔temperature** (`thinkingParamRejection.antagonist.test.ts`) — full per-provider matrix end-to-end: Anthropic/OpenAI-non-none/xAI reject, Gemini accepts, OpenAI `'none'` re-enables, unknown ignores; tier+thinking composition.
5. **Streaming drift + SSE** (`streamingDrift.antagonist.test.ts`) — cross-stream dedup independence; split-mid-keyword SSE reassembly. Guards a module-level (vs per-call) dedup set.
6. **Client-tool continuation** (`clientToolContinuationBuilder.antagonist.test.ts`) — extra fields dropped for Anthropic/Gemini, verbatim for OpenAI/xAI; non-object entries skipped. Guards misapplying the verbatim rule.

## Review

Internal `code-reviewer` pass on the diff. Verdict: suite genuinely adversarial, no over-mocking (reuses existing `fetch`/SSE fixtures), each named wrong-impl actually caught. Findings resolved:
- **P1** — three `as any` casts (repo forbids `any` even in tests) → `as unknown as ReadonlyArray<JsonObject>`; one `as any` eliminated by the class-7 rewrite.
- **P2** — one imperative assertion → declarative `toFailWith` with literal per-case regexes (avoiding the `no-unsafe-regexp` rule a dynamic `RegExp` would trip).
- **P3 (advisory, dispositioned)** — `converters` docblock says "every exported converter" while three trivial enum converters (no optional fields to drop) are not exercised (accurate risk: none); a few near-identical single-value tests could collapse to `test.each` (cosmetic). Left as-is.

## Gates

`rushx build` ✅ · `rushx lint` ✅ (clean, `fixlint` run) · `rushx test` ✅ **100% coverage** (all four metrics; `clientToolContinuationBuilder.ts` with the guard at 100%). Changeset: `patch` (the production guard). Public API surface unchanged (`hasGeminiToolConflict` is `@internal`; `api.md` unchanged).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
_Generated by [Claude Code](https://claude.ai/code/session_01T4rVgY8NFE7zNBNLgEVCch)_

## Summary by CodeRabbit

* **Bug Fixes**
  * Added a clearer pre-check for an unsupported Gemini tool combination, so invalid requests fail fast with a helpful message instead of surfacing a generic error.
  * Improved handling of provider responses and streaming edge cases, including safer event parsing and more reliable truncation/finish-status detection.

* **Tests**
  * Expanded adversarial coverage for model routing, tool handling, parameter validation, response conversion, and streaming behavior to reduce regressions.

### CODE REVIEW

**Comment by ErikFortune** [comment]
@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_

File: libraries/ts-extras/src/test/unit/ai-assist/streamingDrift.antagonist.test.ts [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+      apiKey: 'sk',
+      ...TEST_PROMPT.toRequest(),
+      tools: OPENAI_TOOLS,
+      logger: loggerA
+    });
+    expect(resultA).toSucceed();
+    if (resultA.isSuccess()) {
+      await collect(resultA.value);
+    }
+    expect(loggerA.entries.filter((e) => e.level === 'warning')).toHaveLength(1);
```
[the remark this answers is not in the document]
Comment (ErikFortune) [reply libraries/ts-extras/src/test/unit/ai-assist/streamingDrift.antagonist.test.ts:141]: Declining this one on correctness grounds. `toSucceedAndSatisfy`'s callback is synchronous — `predicate.ts` types it `(value: T) => boolean | void` and invokes it as `cb(received.value)` without awaiting. The `value` here is an `AsyncIterable`, and `await collect(stream)` must complete before the next line (`expect(loggerA.entries.filter(warning)).toHaveLength(1)`) because draining the stream is what fires the drift warnings. Moving the drain into an `async` satisfy callback would leave that promise unawaited, so the length assertion would run before the stream drains — making the test incorrect/flaky rather than cleaner.

The narrow-then-await (`expect(x).toSucceed()` + `if (x.isSuccess()) await collect(x.value)`) is the established convention for draining async-iterable `Result`s in this suite (mirrors `streamingAdaptersDriftInstrument.test.ts`). Keeping it as-is.

---
_Generated by [Claude Code](https://claude.ai/code)_
